x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/scheduling/TransformScheduler.java

```diff
@@ -0,0 +1,230 @@
+     */
+    public interface Listener {
+        /**
+         * Called whenever scheduled task is being triggered.
+         *
+         * @param event event structure containing data that might be useful to the listener
+         */
+        void triggered(Event event);
+    }
+
+    private static final Logger logger = LogManager.getLogger(TransformScheduler.class);
+
+    private final Clock clock;
+    private final ThreadPool threadPool;
+    private final TimeValue schedulerFrequency;
+    private final TransformScheduledTaskQueue scheduledTasks;
+    /**
+     * Prevents two concurrent invocations of the "processScheduledTasks" method.
+     *
+     * Set to {@code true} before processing
+     * Set to {@code false} after processing (doesn't matter whether successful or not).
+     */
+    private final AtomicBoolean isProcessingActive;
+
+    public TransformScheduler(Clock clock, ThreadPool threadPool, Settings settings) {
+        this.clock = Objects.requireNonNull(clock);
+        this.threadPool = Objects.requireNonNull(threadPool);
+        this.schedulerFrequency = Transform.SCHEDULER_FREQUENCY.get(settings);
+        this.scheduledTasks = new TransformScheduledTaskQueue();
+        this.isProcessingActive = new AtomicBoolean();
+    }
+
+    /**
+     * Starts the processing.
+     * Method {@code processScheduledTasks} is invoked according to its configured {@code schedulerFrequency}.
+     */
+    public void start() {
+        threadPool.scheduleWithFixedDelay(this::processScheduledTasks, schedulerFrequency, ThreadPool.Names.GENERIC);
+    }
+
+    // Visible for testing
+    void processScheduledTasks() {
+        // Prevent two concurrent invocations of the "processScheduledTasks" method
+        if (isProcessingActive.compareAndSet(false, true) == false) {
+            return;
+        }
+        logger.trace("Processing scheduled tasks started");
+        Instant processingStarted = clock.instant();
+        try {
+            processScheduledTasksInternal();
+        } finally {
+            // Make sure we clear the "isProcessingActive" bit regardless of whether there was a success or failure.
+            // Otherwise, the processing would be stuck forever.
+            isProcessingActive.set(false);
+        }
+        Instant processingFinished = clock.instant();
+        logger.trace(
+            () -> new ParameterizedMessage(
+                "Processing scheduled tasks finished, took {}ms",
+                Duration.between(processingStarted, processingFinished).toMillis()
+            )
+        );
+    }
+
+    private void processScheduledTasksInternal() {
+        if (scheduledTasks.isEmpty()) {
+            // There are no scheduled tasks, hence, nothing to do
+            return;
+        }
+        long currentTimeMillis = clock.millis();
+        TransformScheduledTask scheduledTask = scheduledTasks.first();
```

Comment:
Why don't we pop them/update all the ones that need to fire within a given window? Maybe I am misread things, but it seems we will only ever fire one transform per scheduler iteration. Seems weird to me.

Reply:
Or maybe have a batch size limit if we are worried about a flurry of them... but regardless, just doing one at a time here seems like a weird thing to me

Reply:
This is good comment. The same issue was brought up by Hendrik in the other comment.
This is now fixed so the scheduler processes all the tasks until the current time.

Reply:
woot, thanks! I haven't reviewed the rest. Consider this a one off comment, so no need for me to approve the whole PR.